Fix error in makeScratch

Lythenas opened the pull request:
This fixes the below error(s). The errors are (somtimes) caused when you move a window to the scratch layer.

The following worked mostly consistently to reproduce this:

- Empty workspace
- Press `<Super>`
- Open gnome-terminal (i.e. "Terminal") by typing "Terminal" and pressing enter
- Press `<Super>`
- Open a second gnome-terminal by typing "Terminal" and pressing `<Ctrl>+<Enter>`
- Move the terminal to the scratch layer by pressing `<Ctrl>+<Super>+<Escape>`
- See error in `journalctl -ef _COMM=gnome-shell`

This error didn't cause big actual errors (besides the error message). And usually PaperWM still behaved completely normally. But sometimes the workspace kind of broke.

<details>
<summary>Errors</summary>

```
JS ERROR: TypeError: this[index] is undefined
moveDone@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:1016:13
_makeEaseCallback/<@resource:///org/gnome/shell/ui/environment.js:153:13
_easeActor@resource:///org/gnome/shell/ui/environment.js:242:17
init/Clutter.Actor.prototype.ease@resource:///org/gnome/shell/ui/environment.js:365:19
addTween@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/utils.js:461:15
layoutColumnSimple@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:444:29
layout@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:524:26
removeWindow@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:745:14
remove_handler@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:2852:11
dynamic_function_ref/<@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/utils.js:104:33
makeScratch@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/scratch.js:77:16
toggle@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/scratch.js:139:20
dynamic_function_ref/<@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/utils.js:104:33
asKeyHandler/<@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/keybindings.js:271:20
```

```
moveDone@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:1021:22
_makeEaseCallback/<@resource:///org/gnome/shell/ui/environment.js:153:13
_easeActor@resource:///org/gnome/shell/ui/environment.js:242:17
init/Clutter.Actor.prototype.ease@resource:///org/gnome/shell/ui/environment.js:365:19
addTween@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/utils.js:461:15
layout@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:558:21
removeWindow@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:745:14
remove_handler@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/tiling.js:2784:11
dynamic_function_ref/<@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/utils.js:104:33
makeScratch@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/scratch.js:77:16
toggle@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/scratch.js:139:20
dynamic_function_ref/<@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/utils.js:104:33
asKeyHandler/<@/home/ms/.local/share/gnome-shell/extensions/paperwm@hedning:matrix.org/keybindings.js:271:20
```

The errors are slightly different, second one does not call `layoutColumnSimple`, but both ultimately end in the same place.

</details>

Notes on the implementation: `ensureViewport()` set `this.selectedWindow` correctly and `this.layout()` needs this to be set correctly. Invoking layout after it also should not cause an other issues because the window was removed and therefore `ensureViewport` wouldn't move the viewport anyway.

TODO: Still need to test on multimonitor and what happens when moving a window to another monitor.

jtaala commented:
Again, nice pickup @Lythenas!

Explanation makes sense to me.  Let me know how your testing goes.

jtaala reviewed (approved):
Explanation makes sense. Work to reproduce and test appreciated.  Approved.

jtaala commented on tiling.js:
```diff
@@ -742,12 +742,12 @@ var Space = class Space extends Array {
         } else {
             this.selectedWindow = null;
         }
+        this.layout();
```
Hmmm, I was thinking of this change while driving today.  Although moving layout to the end for this use case might solve that issue - it might cause issues for other use cases.

Basically, the normal approach is "layout to configure all the tiled windows" and then "ensureViewport to bring the selected window into view", which also sets the tracked `seletectedWindows` variable.

Anyways, what I'm trying to say is that instead of changing the layout() position to be in an unnatural position (e.g. ensuring viewport then layout doesn't really make sense), it would seem better to set `this.selectedWindow` to `selected` if it exists before calling `layout()`.

Will test this out a bit.  Although I'm having trouble reproducing the error in the first place (just checking, are you on X11 or Wayland?).

jtaala replied:
Will test this approach a bit and compare against just moving layout().

jtaala replied:
P.S. I may be overthinking things...

jtaala commented:
> because the window was removed and therefore `ensureViewport` wouldn't move the viewport anyway.

Close, but https://github.com/paperwm/PaperWM/blob/999ad94c45c4d9ec3c8a5da9c3d6d1b8831f5c66/tiling.js#L718-L726
changes `selected` to another window (last most recent one if selected is the window to be removed).

jtaala commented:
@Lythenas I need some help in reproducing this.  I'm on `develop` branch and tried a bunch following precisely your instructions but haven't been able to reproduce the errors.

Any ideas?  Can you get to a reproducible state and run `gather-system-info.sh` for me?  Want to replicate what you have.

jtaala committed:
Changed to work with this.selectedWindow directly (previous implemented
didn't actually update this.selectedWindow before calling layout (e.g.
I believe it should be updated).

jtaala commented:
So, although I haven't been able to reproduce it, after reviewing the code I believe it should be working with (or at least update) `this.selectedWindow` before calling `layout()`.

Previous implementation did this:

- set `selected` to the space's `selectedWindow`;
- check if that window is the one to be removed (which it usually would be);
- if it is, the set `selected` to another window (last selected window in stack of space);
- run `layout`
- run `ensureViewport`

Problem is, the prev implementation didn't actually update `this.selectedWindow`, which I think it should (as you pointed out `this.layout()` expects `this.selectedWindow` to be correct).

I've created [a branch here](https://github.com/paperwm/PaperWM/tree/fix-reported-errors-in-removeWindow) which has changes to remedy this and  work with `this.selectedWindow` instead (which I think is more appropriate).

Can you test this branch out and see if you can reproduce the errors on your machine?

P.S. I also noted your [fix-makeScratch](https://github.com/Lythenas/PaperWM/tree/fix-makeScratch) branch is not actually based on `develop` - it's based on tag `v44.1.1` there's been a few changes/fixes in since then so prob best to update that.

Thanks!

jtaala committed:
Added comment for reason to keep else this.selectedWindow = null line.

jtaala commented:
P.S. the approach outlined above makes sense to me, since to reproduce you needed two windows (which meant that `selected` did exist... and had `layout()` known to use that, it shouldn't have errored out... that's my hypothesis anyway :thinking:).

Lythenas commented:
> Any ideas? Can you get to a reproducible state and run `gather-system-info.sh` for me? Want to replicate what you have.

I can reproduce it quite consistently. On a VM with Ubuntu 23.04 (Gnome 44, wayland). Simply by following the steps in the PR description. Actually I noticed that you don't even need a second window. The error also prints when you move the first/only window to the scratch layer.

If it helps:
```
Distribution: Ubuntu
GNOME Shell 44.1
Display server: Wayland
PaperWM branch/tag: fix-makeScratch
PaperWM commit: 3b7f636758368b606eaaa8fd1882be8e0f7d2deb
Enabled extensions:
- paperwm@hedning:matrix.org
- switcher@landau.fi
- ubuntu-appindicators@ubuntu.com
```

What you explain makes sense, and your branch fix-reported-errors-in-removeWindow also fixes the errors for me. Do you want to open a PR from your branch then?

jtaala commented:
> What you explain makes sense, and your branch fix-reported-errors-in-removeWindow also fixes the errors for me. Do you want to open a PR from your branch then?

I can do that, otherwise you can merge that branch into this branch/PR and I can approve.  Either way works.

jtaala commented:
> > What you explain makes sense, and your branch fix-reported-errors-in-removeWindow also fixes the errors for me. Do you want to open a PR from your branch then?
> 
> I can do that, otherwise you can merge that branch into this branch/PR and I can approve. Either way works.

Alternatively, I can implement this fix into #546 and merge it in in one go.  Let me know what you think and we'll do that.

Lythenas commented:
I pushed your commits to this branch to make it a bit easier for you.
